Index: SDL_gfx-2.0.17/Makefile.am
===================================================================
--- SDL_gfx-2.0.17.orig/Makefile.am	2008-08-02 17:54:04.000000000 +0200
+++ SDL_gfx-2.0.17/Makefile.am	2008-08-02 17:54:44.000000000 +0200
@@ -20,7 +20,7 @@
 
 DEFS = @DEFS@ -DBUILD_DLL
 libSDL_gfx_la_LDFLAGS = 	\
-	-version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
+	-no-undefined -version-info $(LT_CURRENT):$(LT_REVISION):$(LT_AGE)
 
 DISTCLEANFILES = *~ *~c *~h *.cross.cache inc 
 
Index: SDL_gfx-2.0.17/README
===================================================================
--- SDL_gfx-2.0.17.orig/README	2008-08-02 17:54:49.000000000 +0200
+++ SDL_gfx-2.0.17/README	2008-08-02 17:58:54.000000000 +0200
@@ -242,10 +242,14 @@
 Cross-Compilation
 -----------------
 
-The following sequence may be required to cross-compile a DLL:
-	rm -f acinclude.m4
-	sed -i -e 's/-version-info/-no-undefined -version-info/' Makefile.am
-	autoreconf --force --install
+Specify the path to your cross-compiled 'sdl-config', and invoke
+'./configure' with the '--host' and '--build' arguments. For example,
+to cross-compile a .DLL from GNU/Linux:
+
+SDL_CONFIG=/usr/local/cross-tools/i386-mingw32msvc/bin/sdl-config \
+  ./configure --host=i586-mingw32msvc --build=i686-pc-linux-gnu
+make
+make install
 
 No-MMX
 ------
