--- u_canonical	2008-07-15 17:09:13.000000000 +0200
+++ u2_canonical	2008-07-15 17:09:19.000000000 +0200
@@ -24,6 +24,10 @@
       get_frame = true;
     }
   check_joystick ();
+//redink1 Only accept keyboard input when console is not active.
+
+  if (mConsoleActive == false)
+    {
   if ((GetKeyboard (77)) && (GetKeyboard (18)))
     {
 //shutdown music
@@ -43,6 +47,7 @@
 	  debug_mode = true;
 	}
     }
+    }
   if ((GetKeyboard (81)) && (GetKeyboard (18)))
     {
 //shutdown game
@@ -107,6 +145,10 @@
     regcode = mycode;
     sound_support = mcc;
   }
+  if (mLoopMidi && !nothing_playing () && strlen (last_midi) > 0)
+    {
+      PlayMidi (last_midi);
+    }
   if (showb.active)
     {
 //grab main loop and divert it to show a bmp instead
@@ -405,7 +449,8 @@
 				spr[h].seq = 0;
 				spr[h].nocontrol = false;
 				if (h == 1)
-				  if (in_this_base (spr[h].seq_orig, 310))
+				  if (in_this_base
+				      (spr[h].seq_orig, mDinkBasePush))
 				    {
 				      play.push_active = false;
 				      if (play.push_dir == 2)
@@ -494,7 +539,8 @@
 				  spr[h].seq = 0;
 				  spr[h].nocontrol = false;
 				  if (h == 1)
-				    if (in_this_base (spr[h].seq_orig, 310))
+				    if (in_this_base
+					(spr[h].seq_orig, mDinkBasePush))
 				      {
 					play.push_active = false;
 					if (play.push_dir == 2)
@@ -758,6 +804,31 @@
 	      DrawText (hdc, msg, lstrlen (msg), &rcRect, DT_WORDBREAK);
 	    }
 	}
+// Console
+
+      if (mConsoleActive == true)
+	{
+	  SetBkMode (hdc, OPAQUE);
+	  RECT lConsoleRectangle;
+	  lConsoleRectangle.top = 380;
+	  lConsoleRectangle.left = 20;
+	  lConsoleRectangle.right = 620;
+	  lConsoleRectangle.bottom = 400;
+	  char lLine[255];
+	  strncpy (lLine, mConsoleLine.c_str (), 254);
+	  replace ("&", "~~", lLine);
+	  replace ("~~", "&&", lLine);
+	  SetBkColor (hdc, RGB (255, 255, 255));
+	  SetTextColor (hdc, RGB (0, 0, 0));
+	  DrawText (hdc, lLine, lstrlen (lLine), &lConsoleRectangle,
+		    DT_WORDBREAK);
+	  lConsoleRectangle.top = 360;
+	  lConsoleRectangle.bottom = 380;
+	  sprintf (lLine, "%d", mConsoleReturnValue);
+	  SetTextColor (hdc, RGB (255, 0, 0));
+	  DrawText (hdc, lLine, lstrlen (lLine), &lConsoleRectangle,
+		    DT_WORDBREAK);
+	}
       SelectObject (hdc, hfont_small);
       SetBkMode (hdc, TRANSPARENT);
       for (j = 1; j < max_s + 1; j++)
